omniplanner/src/dsg_pddl/dsg_pddl_grounding_Multirobot_Better.py

```diff
@@ -0,0 +1,516 @@
+    filtered_goal_string = filter_goal_for_available_objects(
+        raw_pddl_goal_string, available_object_names
+    )
+    print(f"DEBUG: Filtered goal result: {filtered_goal_string}")
+
+    logger.info(f"Original goal: {raw_pddl_goal_string}")
+    logger.info(f"Filtered goal: {filtered_goal_string}")
+    logger.info(f"Available objects: {available_object_names}")
+
+    # Selected symbol list for connectivity
+    selected_symbols: List[PddlSymbol] = place_symbols + object_symbols
+    # Ensure robots are present in the symbols map for downstream planners
+    for rid, pose in robot_states.items():
+        selected_symbols.append(
+            PddlSymbol(rid, "robot", [], position=np.array(pose[:2]))
+        )
+
+    # Objects section
+    robot_ids = list(robot_states.keys())
+
+    # Init facts
+    connectivity = generate_dense_symbol_connectivity(G, selected_symbols)
+    connected_facts = symbol_connectivity_to_pddl(connectivity)
+
+    # Map objects to nearest place (object-at)
+    object_at_facts: List[Tuple] = []
+    for obj in object_symbols:
+        nearest_place = nearest_place_for_position(place_symbols, obj.position)
+        object_at_facts.append(("object-at", obj.symbol, nearest_place))
+
+    # Suspicious all objects by default
+    suspicious_facts = [("suspicious", o.symbol) for o in object_symbols]
+
+    # Robot starts at nearest places to their given 2D states
+    at_poi_facts = []
+    for rid, pose in robot_states.items():
+        nearest_place = nearest_place_for_position(place_symbols, np.array(pose[:2]))
+        at_poi_facts.append(("at-poi", rid, nearest_place))
+
+    # Compose PDDL
+    def fmt_objects_line(names: List[str], typ: str) -> str:
+        if not names:
+            return ""
+        return " ".join(names) + f" - {typ}\n"
+
+    objects_section = ""
+    objects_section += fmt_objects_line(robot_ids, "robot")
+    objects_section += fmt_objects_line([p.symbol for p in place_symbols], "place")
+    objects_section += fmt_objects_line(
+        [o.symbol for o in object_symbols], "dsg_object"
+    )
+
+    # Serialize init facts
+    def fmt_sexpr(expr: Any) -> str:
+        if isinstance(expr, tuple):
+            head = expr[0]
+            args = expr[1:]
+            return f"({head} {' '.join(fmt_sexpr(a) for a in args)})"
+        return str(expr)
+
+    def fmt_fact(f: Tuple) -> str:
+        head = f[0]
+        if head == "=":
+            # Expect (= (<function> ...) <value>)
+            return f"(= {fmt_sexpr(f[1])} {fmt_sexpr(f[2])})"
+        return f"({head} {' '.join(fmt_sexpr(x) for x in f[1:])})"
+
+    init_facts: List[str] = []
+    # initialize total-cost
+    init_facts.append("(= (total-cost) 0)")
+    init_facts += [fmt_fact(f) for f in at_poi_facts]
+    init_facts += [fmt_fact(f) for f in connected_facts]
+    init_facts += [fmt_fact(f) for f in object_at_facts]
+    init_facts += [fmt_fact(f) for f in suspicious_facts]
+
+    problem_str = """
```

Comment:
Take a look at `PddlProblem` in `pddl_grounding.py` -- We have a class where you can set `domain`, `objects`, `initial_facts`, `goal`, and turn optimizing on or off, and it will dump to the correct PDDL problem string